### PR TITLE
add reference fields to query filter

### LINKED ISSUE #176: Datastore - FindAll(example, pageable) ignores @Reference annotated fields
see https://github.com/spring-cloud/spring-cloud-gcp/issues/2611

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #176

### CODE REVIEW

File: spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java [hunk 1]
```diff
@@ -818,19 +817,30 @@ private DatastoreReaderWriter getDatastoreReadWriter() {
...
 		LinkedList<StructuredQuery.Filter> filters = new LinkedList<>();
+		NullHandler nullHandler = example.getMatcher().getNullHandler();
 		persistentEntity.doWithColumnBackedProperties((persistentProperty) -> {
 			if (!ignoredProperty(example, persistentProperty)) {
 				Value<?> value = getValue(example, probeEntity, persistentEntity, persistentProperty);
-				NullHandler nullHandler = example.getMatcher().getNullHandler();
 				addFilter(nullHandler, filters, persistentProperty.getFieldName(), value);
 			}
 		});
+		persistentEntity.doWithAssociations((AssociationHandler<DatastorePersistentProperty>) association -> {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java:828]: Unnecessary cast?

File: spring-cloud-gcp-data-datastore/src/test/java/com/google/cloud/spring/data/datastore/it/DatastoreIntegrationTests.java [hunk 2]
```diff
@@ -177,6 +185,26 @@ public void saveEntities() {
 
 	}
 
+	@Test
+	public void testFindByExampleReference() {
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/test/java/com/google/cloud/spring/data/datastore/it/DatastoreIntegrationTests.java:189]: Would it make sense to add another test for when the reference is `null`?

**Review by meltsufin — APPROVED** [review]
Looks good overall.
Just two optional comments.

File: spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java (hunk 1 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java:828]: this is necessary, otherwise there is ambiguous call error

File: spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java [hunk 3]
```diff
@@ -818,19 +817,30 @@ private DatastoreReaderWriter getDatastoreReadWriter() {
...
+						this.datastoreMappingContext.getPersistentEntity(value.getClass()));
+				addFilter(nullHandler, filters, property.getFieldName(), KeyValue.of(key));
+			}
+		});
+
+		StructuredQuery.Builder<?> builder =
+				keyQuery
+						? Query.newKeyQueryBuilder()
+						: Query.newEntityQueryBuilder();
+		builder.setKind(persistentEntity.kindName());
```
Comment (elefeint) [inline spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java:843]: This block used to be done outside of the callback. Did you move it into the callback for readability? Just curious.

File: spring-cloud-gcp-data-datastore/src/test/java/com/google/cloud/spring/data/datastore/entities/Product.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		if (this == o) {
+			return true;
+		}
+		if (!(o instanceof Product)) {
+			return false;
+		}
+
+		Product product = (Product) o;
+
+		if (id != null ? !id.equals(product.id) : product.id != null) {
```
Comment (elefeint) [inline spring-cloud-gcp-data-datastore/src/test/java/com/google/cloud/spring/data/datastore/entities/Product.java:54]: Was this autogenerated? The double negatives are making my brain hurt, and I speak Russian!

**Review by elefeint — APPROVED** [review]
Couple of theoretical questions, but actually lgtm.

File: spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java (hunk 3 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/main/java/com/google/cloud/spring/data/datastore/core/DatastoreTemplate.java:843]: yes, just to improve readability

File: spring-cloud-gcp-data-datastore/src/test/java/com/google/cloud/spring/data/datastore/entities/Product.java (hunk 4 above)
Comment (dmitry-s) [reply spring-cloud-gcp-data-datastore/src/test/java/com/google/cloud/spring/data/datastore/entities/Product.java:54]: you guessed it right, it is autogenerated!
